### PR TITLE
Add Starlark Support

### PR DESCRIPTION
**Pull request by clayscode**
This PR adds initial Starlark support into Velociraptor. It adds two functions, `starl_func` and `starl_vars`. `starl_func` allows users to compile and reuse starlark functions in VQL. `starl_vars` allows for the retrieval of variables within the compiled starlark code global namespace. Example usages:
1) Retrieve a variable with starl_vars
```yaml
name: Custom.StarlarkExample
description: Example Starlark Execution

parameters:
  - name: CodeBlock
    default: |
      def ps_enum(ps_list):
        return [x.Pid for x in ps_list]
      a = 3
sources:
  - queries:
      - SELECT starl_vars(code=CodeBlock, vars=["a"]) FROM scope()
```
Output: 
```json
[
 {
  "starl_vars(code=CodeBlock, vars= [\"a\"])": {
   "a": 3
  }
 }
]
```
2) Call a function with starl_func
```yaml
name: Custom.StarlarkExample
description: Example Starlark Execution

parameters:
  - name: CodeBlock
    default: |
      def ps_enum(ps_list):
        return [x.Pid for x in ps_list]
      a = 3
sources:
  - queries:
      - LET ps_enum(input_list) = starl_func(code=CodeBlock, func="ps_enum", ps_list=input_list)
      - LET pids <= SELECT * from pslist() LIMIT 5
      - SELECT ps_enum(input_list=pids) FROM scope()
```
Output:
```json
[
 {
  "ps_enum(input_list=pids)": [
   1,
   2,
   3,
   4,
   6
  ]
 }
]
```
3) Pass globals to Starlark and retrieve results
```yaml
name: Custom.StarlarkExample
description: Example Starlark Execution

parameters:
  - name: CodeBlock
    default: |
      def ps_enum(ps_list):
        return [x.Pid for x in ps_list]
      a = 3 * b
sources:
  - queries:
      - SELECT starl_vars(code=CodeBlock, vars=["a"], globals=dict(b=3)) FROM scope()
```
Output
```json
[
 {
  "starl_vars(code=CodeBlock, vars= [\"a\"], globals=dict(b=3))": {
   "a": 9
  }
 }
]
```

Note that starl_func can also kwargs, args, or nothing:
```sql
LET ps_enum(input_list) = starl_func(code=CodeBlock, func="ps_enum", kwargs=dict(ps_list=input_list))
LET ps_enum(input_list) = starl_func(code=CodeBlock, func="ps_enum", args=[input_list])
LET ps_enum(input_list) = starl_func(code=CodeBlock, func="ps_enum")
```

### CODE REVIEW

**Comment by clayscode** [comment]
Hmm, just realized this doesn't work if you try to use multiple code blocks without specifying a new key. Perhaps the cache key should just be a hash of the code block?

**Comment by clayscode** [comment]
Also just noticed I forgot to add support for starlark.Tuple type in `starlarkValueAsInterface`. Will have to patch that in tomorrow.

File: vql/tools/starlark.go [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+	}
+
+	return compiled_vars, nil
+}
+
+// User doesn't actually specify extras, that's done in extractStarlarkFuncArgs
+type StarlarkFuncArgs struct {
+	Func    string      `vfilter:"required,field=func,doc=Starlark function to call."`
+	Code    string      `vfilter:"required,field=code,doc=Starlark code to compile."`
+	Args    vfilter.Any `vfilter:"optional,field=args,doc=Positional args for the function."`
```
Comment (scudette) [inline vql/tools/starlark.go:269]: You probably want a LazyExpr here. The only time you need an Any is if you want to do the materializing yourself. Any literally does not do any conversion or type checking so you have to do a lot of it by hand.

A LazyExpr will be available when you call Reduce()

File: vql/tools/starlark.go [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+
+	return starlark.Float(floatValue), nil
+}
+
+// Get scope of compiled starlark context
+func getCompiled(ctx context.Context,
+	scope vfilter.Scope,
+	key string, code string, globals vfilter.Any) (*ordereddict.Dict, error) {
+	if key == "" {
+		new_hash := sha256.New()
```
[the remark this answers is not in the document]
Comment (clayscode) [reply vql/tools/starlark.go:247]: True, was just worried about using too much memory but guess that's probably a silly concern. The program would have to be truly massive for that to be a problem and it wouldn't be the code block but the compiled values themselves.

File: vql/tools/starlark.go [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+	sthread := &starlark.Thread{Name: "VQL Thread"}
+	predeclared, err := interfaceAsStarlarkValue(globals)
+	if err != nil {
+		return err
+
+	}
+	stringdict := starlark.StringDict{}
+	switch predeclared.(type) {
+	case *starlarkstruct.Struct:
+		predeclared.(*starlarkstruct.Struct).ToStringDict(stringdict)
```
[the remark this answers is not in the document]
Comment (clayscode) [reply vql/tools/starlark.go:288]: ah sorry, meant to clean that up

File: vql/tools/starlark.go [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+
+// Turn dicts into starlark tuples to pass to starlark.Call
+func makeKwargsTuple(kwargs *ordereddict.Dict) ([]starlark.Tuple, error) {
+	starlark_args, err := interfaceAsStarlarkValue(kwargs)
+	if err != nil {
+		return nil, err
+	}
+	starlark_tuple := make([]starlark.Tuple, 0)
+	new_string_dict := starlark.StringDict{}
+	starlark_args.(*starlarkstruct.Struct).ToStringDict(new_string_dict)
```
[the remark this answers is not in the document]
Comment (clayscode) [reply vql/tools/starlark.go:336]: Since kwargs will always be a dict, starlark_args should always be a struct. Fair point though, will add a check.

**Pull request merged**